src/digraphs.c

```diff
@@ -1384,6 +1384,80 @@ static bool EqJumbledPlists(Obj l, Obj r, Int nr, Int* buf) {
   return true;
 }
 
+//-----------------------------------------------------------------------------
+// MurmurHash3 was written by Austin Appleby, and is placed in the public
+// domain. The author hereby disclaims copyright to this source code.
+
+/* Minor modifications to get it to compile in C rather than C++ and
+integrate with GAP  SL*/
+
+/* Digraphs takes parts of this source from GAP */
+
+#define BIG_CONSTANT(x) (x##LLU)
+
+#ifndef SYS_IS_64_BIT
+
+static inline uint32_t fmix4(uint32_t h) {
+  h ^= h >> 16;
+  h *= 0x85ebca6b;
+  h ^= h >> 13;
+  h *= 0xc2b2ae35;
+  h ^= h >> 16;
+
+  return h;
+}
+
+#else
+
+static inline uint64_t fmix8(uint64_t k) {
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xff51afd7ed558ccd);
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xc4ceb9fe1a85ec53);
+  k ^= k >> 33;
+
+  return k;
+}
+
+#endif
+
+static Obj FuncDIGRAPH_HASH(Obj self, Obj digraph) {
+  UInt n, i;
+  Obj  out, a, res;
+  Int  nr, j;
+
+  n   = DigraphNrVertices(digraph);
+  out = FuncOutNeighbours(self, digraph);
+
+#ifndef SYS_IS_64_BIT
+  uint32_t* buf;
+  buf = safe_calloc(n, sizeof(uint32_t));
+#else
+  uint64_t* buf;
+  buf = safe_calloc(n, sizeof(uint64_t));
```

Comment:
Hash functions are supposed to be as fast and low overhead as possible. Doing allocations in them is therefore suboptimal.

In this case, you feed this buffer into `HASHKEY_MEM_NC` in the end, which just has to read everything in that buffer out again and mix it back together.

So you could avoid using a buffer if just mix everything together it, using a "good enough" mixing function -- here by mixing function I mean one which takes two `UInt` and combines them into a single UInt in a way that's "good foro hasing. E.g. it should be far from commutative and associative.

(I think this also referred to as "hash combiner" and possibly other names).

A very simple hash combiner better than `+` would be
```
static UInt hash_combine(UInt a, UInt b) {
  return fmix(a + fmix(b))
}
```
but it is inefficient because it invokes `fmix` twice. I'll see if I can find a web page with better suggestions.

Reply:
E.g. https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine suggests
```C
seed ^= hash_value(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```